pydra/engine/submitter.py

```diff
@@ -110,41 +98,37 @@ async def submit(self, runnable, wait=False, rerun=False):
             Coroutines for :class:`~pydra.engine.core.TaskBase` execution.
 
         """
+        if runnable.plugin and runnable.plugin != self.plugin:
+            raise NotImplementedError()
+
         futures = set()
-        if runnable.state:
-            runnable.state.prepare_states(runnable.inputs, cont_dim=runnable.cont_dim)
-            runnable.state.prepare_inputs()
-            logger.debug(
-                f"Expanding {runnable} into {len(runnable.state.states_val)} states"
-            )
-            task_pkl = runnable.pickle_task()
-
-            for sidx in range(len(runnable.state.states_val)):
-                job_tuple = (sidx, task_pkl, runnable)
-                if is_workflow(runnable):
-                    # job has no state anymore
-                    futures.add(self.submit_workflow(job_tuple, rerun=rerun))
-                else:
-                    # tasks are submitted to worker for execution
-                    futures.add(self.worker.run_el(job_tuple, rerun=rerun))
-        else:
+        if runnable.state is None:
+            raise Exception("Only runnables with state should reach here")
+
+        task_pkl = await prepare_runnable_with_state(runnable)
+
+        for sidx in range(len(runnable.state.states_val)):
             if is_workflow(runnable):
-                await self._run_workflow(runnable, rerun=rerun)
+                # job has no state anymore
+                futures.add(
+                    # This unpickles and runs workflow - why are we pickling?
+                    asyncio.create_task(load_and_run_async(task_pkl, sidx, self, rerun))
```

Comment:
I think this bit is still excessively confusing, and could really benefit from a rework / simplification. `Worker.run_el` should have a streamlined behavior - currently it handles two pathways (state vs stateless)